source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -536,7 +541,14 @@ void HappyEyeballsConnectionImpl::onBelowWriteBufferLowWatermark(
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
-void HappyEyeballsConnectionImpl::onWriteBufferHighWatermark() { ASSERT(!connect_finished_); }
+void HappyEyeballsConnectionImpl::onWriteBufferHighWatermark() {
+  ASSERT(!connect_finished_);
+  for (auto callback : post_connect_state_.connection_callbacks_) {
+    if (callback) {
+      callback->onAboveWriteBufferHighWatermark();
```

Comment:
won't this result in an uneven number of calls if we don't pass through the low watermark callbacks as well?

Reply:
I'm definitely not an expert in this space, so I'm probably wrong. :)  But this method is only called from the deferred write buffer. If we ever go over the highwark, we will deliver this signal (obviously). But then when the connection connections we will move this data to the underlying connection. We won't hit the low water mark until the underlying connection actually writes the data. At that point, the connection will call `onBelowWriteBufferHighWatermark()`. Does that sound right to you?

Reply:
*nod* I added two comments about this. One at the end of `onEvent()` which says:

```
  // Add connection callbacks after moving data from the deferred write buffer so that                                                                                                             
  // any high watermark notification is swallowed and not conveyed to the callbacks, since                                                                                                         
  // that was already delivered to the callbacks when the data was written to the buffer.                                                                                                          
```

And the other in onWriteBufferLowWatermark() 

```
  // Only called when moving write data from the deferred write buffer to                                                                                                                          
  // the underlying connection. In this case, the connection callbacks must                                                                                                                        
  // not be notified since this should be transparent to the callbacks.                                                                                                                            
```
Happy to add more elsewhere if you think that would be helpful?

Reply:
I think this makes sense to me!